### ISSUE TITLE
comma-dangle inside promises is weird

### ISSUE THREAD

**Issue by felixsanz** [eslint/eslint#7786]
Sadly, `.then()` can accept multiple arguments (the second being the `catch` function). But none should really use `.then()` with more than one argument (right? :stuck_out_tongue_closed_eyes: )

In this example eslint complains about comma-dangle rule:

```javascript
y.then(x => x
  .replace('asdasdsdsda', '123213213')
  .replace('cxvcxvcxvcx', '546546546')
  .replace('ewrewrewrew', '877876876'), // <- this is confusing as hell. And useless comma!
)
```

I had problems trying to understand it at first because i wasn't thinking that the comma belongs to the `.then()` arguments. I thought the comma was part of the `replace` chain and it was a bit hard to spot the problem :disappointed: 

A workaround could be:

```javascript
y.then(x => (x
  .replace('asdasdsdsda', '123213213')
  .replace('cxvcxvcxvcx', '546546546')
  .replace('ewrewrewrew', '877876876')
))
```

But this is ugly because the extra parens should be only added when eslint complains to the rule [no-confusing-arrow](http://eslint.org/docs/rules/no-confusing-arrow). I mean, this is not a confusing arrow, so it shouldn't need parens.

I think the comma-dangle rule should be removed from inside `.then()` functions, or at least have an option to relax the rule there, because this seemed like a bug at first.

**Comment by eslintbot** [comment]
Hi @felixsanz, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by ilyavolodin** [comment]
@felixsanz Could you provide the information requested by the bot? Because without it it's very hard to understand what the problem is. Your first example is an invalid syntax that can't be parsed by Espree. You are using arrow shorthand syntax, and comma has nothing to do with `then`, but is part of the `return` expression (which makes it invalid syntax).

**Comment by felixsanz** [comment]
@ilyavolodin That was the first thing I was thinking of. I thought the comma was part of the expression and it was a comma operator but nope! The comma is indeed part of the `then`, take a look:

```javascript
y.then(x => x
  .replace('ewrewrewrew', '877876876'), // <- this is confusing as hell. And useless comma!
  (e) => { console.error(e) }
)
```

You can see this online: https://jsbin.com/lamuginulu/edit?js,console

ESLint version is 3.9.1

**Comment by felixsanz** [comment]
And you saying that this is invalid syntax confirms that this is really weird and hard to reason about

**Comment by mysticatea** [comment]
@felixsanz could you answer the question of https://github.com/eslint/eslint/issues/7786#issuecomment-267741976 ?

**Comment by felixsanz** [comment]
1. `3.9.1`
2. what i did is already on the first message. the eslint configuration is just the standard airbnb style guide
3. i don't know what is this, output?
4. not sure what to answer here, what i expected is already one first message also

does this help?

**Comment by ljharb** [comment]
Here's another workaround that avoids having a multiline arrow function look confusingly like it's a one-liner:
```js
y.then(x => (
  x.replace('asdasdsdsda', '123213213')
    .replace('cxvcxvcxvcx', '546546546')
    .replace('ewrewrewrew', '877876876')
))
```

**Comment by felixsanz** [comment]
@ljharb that's exactly what i said on my first message, on the second example.

You can wrap it with parens (doesn't matter if you put the `x` in the parens line or as you did). But this is not a `no-confusing-arrow` so parens shouldn't be used at all, right?

I think using the second argument of ´then()` should be totally discouraged (thats why `catch()` exists). And this comma here is just so confusing. It's almost a must to have an option to relax the `comma-dongle` rule inside `then()`, or totally remove `comma-dongle` behaviour inside `then` :/

**Comment by ljharb** [comment]
Airbnb's style guide does in fact require those parens (if it's not documented, it should be), because the body of the implicit return arrow function is spread across more than one line.

This has nothing to do with `.then`, because `comma-dangle` would be requiring the comma with any function call - `eslint` has no way of knowing how many arguments the given function tends to take. If you don't want this `comma-dangle` behavior, then you need to turn it off for all function invocations - `.then` isn't relevant.

**Comment by felixsanz** [comment]
@ljharb Here it is: https://github.com/airbnb/javascript#arrows--paren-wrap

I didn't know that (it's hard to remember all the style guide by memory). So it's fine now i think, my fault probably (but it's so confusing if you don't know this recommended parens!)

I'm gonna close this, re-open if neccesary. Thx

**Comment by mysticatea** [comment]
@felixsanz 

- [Airbnb config enables `comma-dangle` rule with `functions: 'always-multiline'` option](https://github.com/airbnb/javascript/blob/e037d29b93e5b61434deb9ba485f76c6406b7cb3/packages/eslint-config-airbnb-base/rules/errors.js#L4). The `always-multiline` option requires dangling commas if the closing parenthesis is not on the same line of the last argument. (This is not default.)
- In your first example, the closing parenthesis is not on the line of the last argument.
- So `comma-dangle` requires the dangling comma.
